### PR TITLE
Stop setting element_fp to zero in defualt constructor, and do not create a new zero instance each time function zero() is called.

### PR DESCRIPTION
**Pull request by martun**

### CODE REVIEW

File: include/nil/crypto3/algebra/fields/field.hpp [hunk 1]
```diff
@@ -70,10 +70,6 @@ namespace nil {
                         nil::crypto3::multiprecision::unchecked, void>
                         modular_backend;
 
-                    typedef nil::crypto3::multiprecision::number<
```
Comment (martun) [inline include/nil/crypto3/algebra/fields/field.hpp:73]: This is never used and wrong, we must not use _rt version anywhere.

**Comment by martun** [comment]
> Quick question. Can't we simply mark the existing function a constexpr one and let the compiler decide when it should be actually evaluated in compile-time?

We cannot have a static variable inside a constexpr function, it does not compile. 

The "correct" way of doing this is through std::is_constant_evaluated, which is a C++20 feature, and clang12 doesn't see it :) 
In case we end up updating all the compilers, we may switch to it and have a nicer code one day.

**Comment by x-mass** [comment]
> The "correct" way of doing this is through std::is_constant_evaluated, which is a C++20 feature, and clang12 doesn't see it :)
In case we end up updating all the compilers, we may switch to it and have a nicer code one day.

I think, this won't work either, since `std::is_constant_evaluated` is evaluated in runtime, and we still can't use `static` variables inside `constexpr`. Works not like `if constexpr`, when another branch is not even compiled

**Comment by martun** [comment]
> > The "correct" way of doing this is through std::is_constant_evaluated, which is a C++20 feature, and clang12 doesn't see it :)
> > In case we end up updating all the compilers, we may switch to it and have a nicer code one day.
> 
> I think, this won't work neither, since `std::is_constant_evaluated` is evaluated in runtime, and we still can't use `static` variables inside `constexpr`. Works not like `if constexpr`, when another branch is not even compiled
> 
> Discussed in dm, that we should try to make it
> 
> ```c++
> static constexpr element_fp zero_value = 0;
> constexpr inline static element_fp zero() {
>     return zero_value;
> }
> ```
> 
> to avoid different functoins for compile- and runtime

This did not compile, since we cannot create a constexpr instance of type element_fp  inside class element_fp , becuase at that line class element_fp  is incomplete type. I moved this outside of the class, into some unnamed namespace, doesn't look nice, but I failed to compile it in any other way.

File: include/nil/crypto3/algebra/fields/detail/element/fp.hpp [hunk 2]
```diff
@@ -64,39 +64,37 @@ namespace nil {
...
-                            : data(std::move(B.data)) {
-                        };
+                            : data(std::move(B.data)) {}
 
-                        constexpr inline static element_fp zero() {
-                            return element_fp(0);
-                        }
+                        // Creating a zero from int is a fairly slow operation. We want all the 
+                        // callers who don't need constexpr to use this version.
+                        inline static element_fp zero();
```
Comment (x-mass) [inline include/nil/crypto3/algebra/fields/detail/element/fp.hpp:86]: These funcs are implicitly inlined, no need to put `inline` keyword

**Review by x-mass — APPROVED** [review]
I like approach with static class fields more than prev one

**Pull request merged**